src/code/z_parameter.c

```diff
@@ -2564,49 +2873,573 @@ s16 D_801BFC40[] = {
+    7 * MINIGAME_PERFECT_ANGLE_PER_LETTER, // E
+    0 * MINIGAME_PERFECT_ANGLE_PER_LETTER, // R
+    1 * MINIGAME_PERFECT_ANGLE_PER_LETTER, // F
+    5 * MINIGAME_PERFECT_ANGLE_PER_LETTER, // E
+    4 * MINIGAME_PERFECT_ANGLE_PER_LETTER, // C
+    3 * MINIGAME_PERFECT_ANGLE_PER_LETTER, // T
+    2 * MINIGAME_PERFECT_ANGLE_PER_LETTER, // !
 };
-s16 D_801BFC60[][3] = {
-    255, 255, 255, 255, 165, 55,
+
+s16 sMinigamePerfectType1PrimColorTargets[2][3] = {
+    { 255, 255, 255 },
+    { 255, 165, 55 },
 };
 
-#pragma GLOBAL_ASM("asm/non_matchings/code/z_parameter/func_8011B5C0.s")
+void Interface_UpdateMinigamePerfectType1(PlayState* play) {
+    InterfaceContext* interfaceCtx = &play->interfaceCtx;
+    s16 i;
+    s16 count;
+    s16 colorStepR;
+    s16 colorStepG;
+    s16 colorStepB;
+
+    // Update letter positions
+    for (count = 0, i = 0; i < interfaceCtx->minigamePerfectLetterCount; i++, count += 4) {
+        if (interfaceCtx->minigamePerfectState[i] == MINIGAME_PERFECT_STATE_INIT) {
+            // Initialize letter positions along the parameteric ellipse curve
+            interfaceCtx->minigamePerfectLetterEllipseAngle[i] = sMinigamePerfectType1AngleOffScreenTargets[i] + 0xA000;
+            interfaceCtx->minigamePerfectState[i] = MINIGAME_PERFECT_STATE_ENTER;
+        } else if (interfaceCtx->minigamePerfectState[i] == MINIGAME_PERFECT_STATE_ENTER) {
+            // Swirl inwards along a swirlic parametric ellipse equation to the spelt-out word
+            interfaceCtx->minigamePerfectLetterEllipseAngle[i] -= 0x800;
+            if (interfaceCtx->minigamePerfectLetterEllipseAngle[i] == sMinigamePerfectType1AngleOffScreenTargets[i]) {
+                interfaceCtx->minigamePerfectState[i] = MINIGAME_PERFECT_STATE_STATIONARY;
+            }
+        } else if (interfaceCtx->minigamePerfectState[i] == MINIGAME_PERFECT_STATE_SPREAD) {
+            // Swirl outwards along a swirlic parametric ellipse equation offscreen
+            interfaceCtx->minigamePerfectLetterEllipseAngle[i] -= 0x800;
+            if (interfaceCtx->minigamePerfectLetterEllipseAngle[i] ==
+                (u16)(sMinigamePerfectType1AngleOffScreenTargets[i] - 0x8000)) {
+                interfaceCtx->minigamePerfectState[i] = MINIGAME_PERFECT_STATE_OFF;
+            }
+        }
+    }
 
-s16 D_801BFC6C[] = {
-    78, 54, 29, 5, -18, -42, -67, -85,
+    // Initialize the next letter in the list
+    // if `minigamePerfectLetterCount == MINIGAME_PERFECT_NUM_LETTERS`,
+    // then minigamePerfectState[] is accessed
+    if ((interfaceCtx->minigamePerfectState[interfaceCtx->minigamePerfectLetterCount] == MINIGAME_PERFECT_STATE_OFF) &&
+        (interfaceCtx->minigamePerfectLetterCount < MINIGAME_PERFECT_NUM_LETTERS)) {
+        interfaceCtx->minigamePerfectState[interfaceCtx->minigamePerfectLetterCount] = MINIGAME_PERFECT_STATE_INIT;
+
+        interfaceCtx->minigamePerfectLetterEllipseRadiusX[interfaceCtx->minigamePerfectLetterCount] = 140.0f;
```

Comment:
I was also concerned about the name length, but I figured it'd be easier to go long -> short rather than the other way around.

I've also brainstormed dropping the `minigame` out of `minigamePerfect` to help with the length. Would `perfectLetters` be find for some of these? But then you do lose consistency as names like `isMinigamePerfect` doesn't feel right as just `isPerfect`

Reply:
Yeah, I think we have to keep `minigame` and `perfect`, the rest is negotiable.

Reply:
I'm not sure we 100% need to keep  since this more about the `PERFECT` word which technically doesn't need to be used for minigames.

Reply:
I looked through all of `z_parameter` and the whole "minigame-perfect-letters" do seem to be completely decoupled and independent to the actual minigame code, which is interesting. In practice, it makes sense to use at the end of a minigame, but the "perfect-letter" system itself can run top-to-bottom independent of any minigame state or status. So maybe it's worth decoupling them a little more and dropping the `minigame` term? That's a tough one though since most would associate it with minigames, and just `perfect` itself might not be descriptive enough (`perfectLetters` might be fine though)

Reply:
Maybe `perfectWord`? (though `pefectLetters` is fine imo)

Reply:
I've gone and changed it to `perfectLetters` and dropped the `minigame` to see what it's like. I can always revert it if it's not the greatest, but it's there to look at now